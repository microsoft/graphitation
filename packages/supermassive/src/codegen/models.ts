import ts, { factory } from "typescript";
import {
  DocumentNode,
  isTypeDefinitionNode,
  isTypeExtensionNode,
} from "graphql";
import { ASTReducer, visit } from "./typedVisitor";
import { TsCodegenContext } from "./context";
import { createNullableType, createNonNullableType } from "./utilities";

export function generateModels(
  context: TsCodegenContext,
  document: DocumentNode,
): ts.SourceFile {
  return visit(document, createModelsReducer(context)) as ts.SourceFile;
}

type ASTReducerMap = {
  Document: ts.SourceFile;
  SchemaExtension: null;
  ObjectTypeExtension: null;

  NameNode: string;

  NamedType: ts.TypeNode;
  ListType: ts.TypeNode;
  NonNullType: ts.TypeNode;

  Directive: null;

  FieldDefinition: ts.PropertySignature;

  ObjectTypeDefinition: ts.InterfaceDeclaration;
  InputObjectTypeDefinition: ts.InterfaceDeclaration;
  UnionTypeDefinition: ts.TypeAliasDeclaration;
  EnumTypeDefinition: ts.EnumDeclaration;
  EnumValueDefinition: ts.EnumMember;
  InterfaceTypeDefinition: ts.InterfaceDeclaration;
  // ScalarTypeDefinition: ts.TypeAliasDeclaration;

  DirectiveDefinition: null;
};

type ASTReducerFieldMap = {
  Document: {
    definitions: ts.Statement | ts.Statement[];
  };

  NamedType: {
    name: ASTReducerMap["NameNode"];
  };
  ListType: {
    type: ts.TypeNode;
  };
  NonNullType: {
    type: ts.TypeNode;
  };

  FieldDefinition: {
    name: ASTReducerMap["NameNode"];
    type: ts.TypeNode;
  };

  ObjectTypeDefinition: {
    name: ASTReducerMap["NameNode"];
    interfaces: ASTReducerMap["NamedType"];
    fields: ASTReducerMap["FieldDefinition"];
  };
  InputObjectTypeDefinition: {
    name: ASTReducerMap["NameNode"];
    fields: ASTReducerMap["FieldDefinition"];
  };
  UnionTypeDefinition: {
    name: ASTReducerMap["NameNode"];
    types: ts.TypeNode[];
  };
  InterfaceTypeDefinition: {
    name: ASTReducerMap["NameNode"];
  };
  EnumTypeDefinition: {
    name: ASTReducerMap["NameNode"];
    values: ASTReducerMap["EnumValueDefinition"];
  };
  EnumValueDefinition: {
    name: ASTReducerMap["NameNode"];
  };
  // ScalarTypeDefinition: {
  //   name: ASTReducerMap["NameNode"];
  // };
};

function createModelsReducer(
  context: TsCodegenContext,
): ASTReducer<ts.Node | string, ASTReducerMap, ASTReducerFieldMap> {
  return {
    Document: {
      leave(node) {
        const imports = context.getAllModelImportDeclarations() as ts.Statement[];
        const statements = node.definitions;
        return factory.createSourceFile(
          imports.concat(statements.flat()),
          factory.createToken(ts.SyntaxKind.EndOfFileToken),
          ts.NodeFlags.None,
        );
      },
    },
    SchemaExtension: {
      leave(): null {
        return null;
      },
    },
    ObjectTypeExtension: {
      leave(): null {
        return null;
      },
    },

    FieldDefinition: {
      leave(node): ts.PropertySignature {
        return factory.createPropertySignature(
          undefined,
          factory.createIdentifier(node.name),
          undefined,
          node.type,
        );
      },
    },

    ObjectTypeDefinition: {
      leave(node): ts.InterfaceDeclaration {
        const model = context.getDefinedModelType(node.name);
        const interfaces = node.interfaces || [];
        const extendTypes = [context.getBaseModelType()];
        if (model) {
          extendTypes.push(model);
        }
        console.log(extendTypes);
        return factory.createInterfaceDeclaration(
          undefined,
          [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
          factory.createIdentifier(`${node.name}Model`),
          undefined,
          [
            factory.createHeritageClause(
              ts.SyntaxKind.ExtendsKeyword,
              extendTypes.map((type) =>
                factory.createExpressionWithTypeArguments(
                  type.toExpression(),
                  undefined,
                ),
              ),
              // .concat(interfaces),
            ),
          ],
          [
            factory.createPropertySignature(
              undefined,
              "__typeName",
              undefined,
              factory.createLiteralTypeNode(
                factory.createStringLiteral(node.name),
              ),
            ),
            ...(node.fields || []),
          ],
        );
      },
    },

    EnumTypeDefinition: {
      leave(node): ts.EnumDeclaration {
        return factory.createEnumDeclaration(
          undefined,
          [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
          `${node.name}Model`,
          node.values || [],
        );
      },
    },

    EnumValueDefinition: {
      leave(node): ts.EnumMember {
        return factory.createEnumMember(
          node.name,
          factory.createStringLiteral(node.name),
        );
      },
    },

    InterfaceTypeDefinition: {
      leave(node): ts.InterfaceDeclaration {
        const extendTypes = [context.getBaseModelType()];
        return factory.createInterfaceDeclaration(
          undefined,
          [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
          factory.createIdentifier(`${node.name}Model`),
          undefined,
          [
            factory.createHeritageClause(
              ts.SyntaxKind.ExtendsKeyword,
              extendTypes.map((type) =>
                factory.createExpressionWithTypeArguments(
                  type.toExpression(),
                  undefined,
                ),
              ),
            ),
          ],
          [
            factory.createPropertySignature(
              undefined,
              "__typeName",
              undefined,
              factory.createTypeReferenceNode("string"),
            ),
          ],
        );
      },
    },

    NamedType: {
      leave(node): ts.TypeNode {
        return createNullableType(
          context.getModelType(node.name).toTypeReference(),
        );
      },
    },
    ListType: {
      leave({ type }): ts.TypeNode {
        return createNullableType(
          factory.createArrayTypeNode(type as ts.TypeNode),
        );
      },
    },
    NonNullType: {
      leave({ type }): ts.TypeNode {
        return createNonNullableType(type as ts.UnionTypeNode);
      },
    },

    Name: {
      leave(node) {
        return node.value;
      },
    },

    Directive: {
      leave() {
        return null;
      },
    },
  };
}
