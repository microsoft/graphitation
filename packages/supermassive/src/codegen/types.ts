import { DirectiveNode } from "graphql";
import ts from "typescript";

export type DefinitionImport = {
  from: string;
  defs: string[];
  importName: string;
  directive: DirectiveNode;
};

export type DefinitionModel = {
  tsType: string;
  from: string;
  directive: DirectiveNode;
};

export type DefinitionModels = Map<string, DefinitionModel>;
