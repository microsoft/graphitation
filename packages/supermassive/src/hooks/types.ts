import type { OperationDefinitionNode } from "graphql";
import type { ResolveInfo, TotalExecutionResult } from "../types";

interface BaseExecuteHookArgs<ResolveContext> {
  context: ResolveContext;
}

export interface BaseExecuteFieldHookArgs<ResolveContext>
  extends BaseExecuteHookArgs<ResolveContext> {
  resolveInfo: ResolveInfo;
}

export interface PostExecuteFieldHookArgs<ResolveContext, HookContext>
  extends BaseExecuteFieldHookArgs<ResolveContext> {
  hookContext: HookContext;
}

export interface AfterFieldResolveHookArgs<ResolveContext, HookContext>
  extends PostExecuteFieldHookArgs<ResolveContext, HookContext> {
  result?: unknown;
  error?: unknown;
}

export interface AfterFieldCompleteHookArgs<ResolveContext, HookContext>
  extends PostExecuteFieldHookArgs<ResolveContext, HookContext> {
  result?: unknown;
  error?: unknown;
}

export interface BaseExecuteOperationHookArgs<ResolveContext>
  extends BaseExecuteHookArgs<ResolveContext> {
  operation: OperationDefinitionNode;
}

export interface AfterBuildResponseHookArgs<ResolveContext>
  extends BaseExecuteOperationHookArgs<ResolveContext> {
  result: TotalExecutionResult;
}

export interface BeforeSubscriptionEventEmitHookArgs<ResolveContext>
  extends BaseExecuteOperationHookArgs<ResolveContext> {
  eventPayload: unknown;
}

export interface BeforeFieldResolveHook<
  ResolveContext = unknown,
  BeforeHookContext = unknown,
> {
  (args: BaseExecuteFieldHookArgs<ResolveContext>):
    | Promise<BeforeHookContext>
    | BeforeHookContext;
}

export interface AfterFieldResolveHook<
  ResolveContext = unknown,
  BeforeHookContext = unknown,
  AfterHookContext = BeforeHookContext,
> {
  (
    args: AfterFieldResolveHookArgs<ResolveContext, BeforeHookContext>,
  ): AfterHookContext;
}

export interface AfterFieldCompleteHook<
  ResolveContext = unknown,
  AfterHookContext = unknown,
> {
  (args: AfterFieldCompleteHookArgs<ResolveContext, AfterHookContext>): void;
}

export interface AfterBuildResponseHook<ResolveContext = unknown> {
  (args: AfterBuildResponseHookArgs<ResolveContext>): void;
}

export interface BeforeOperationExecuteHook<ResolveContext = unknown> {
  (args: BaseExecuteOperationHookArgs<ResolveContext>): void | Promise<void>;
}

export interface BeforeSubscriptionEventEmitHook<ResolveContext = unknown> {
  (
    args: BeforeSubscriptionEventEmitHookArgs<ResolveContext>,
  ): void | Promise<void>;
}

export interface ExecutionHooks<
  ResolveContext = unknown,
  BeforeHookContext = unknown,
  AfterHookContext = BeforeHookContext,
> {
  beforeOperationExecute?: BeforeOperationExecuteHook<ResolveContext>;
  beforeSubscriptionEventEmit?: BeforeSubscriptionEventEmitHook<ResolveContext>;
  beforeFieldResolve?: BeforeFieldResolveHook<
    ResolveContext,
    BeforeHookContext
  >;
  afterFieldResolve?: AfterFieldResolveHook<
    ResolveContext,
    BeforeHookContext,
    AfterHookContext
  >;
  afterFieldComplete?: AfterFieldCompleteHook<ResolveContext, AfterHookContext>;
  afterBuildResponse?: AfterBuildResponseHook<ResolveContext>;
}
