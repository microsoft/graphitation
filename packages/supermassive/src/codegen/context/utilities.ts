import { factory, isImportSpecifier } from "typescript";

export function createImportDeclaration(
  importNames: string[],
  from: string,
  importAlias?: string,
) {
  return factory.createImportDeclaration(
    undefined,
    undefined,
    factory.createImportClause(
      true,
      undefined,
      factory.createNamedImports(
        importNames.map((importName) =>
          factory.createImportSpecifier(
            importAlias ? factory.createIdentifier(importAlias) : undefined,
            factory.createIdentifier(importName),
          ),
        ),
      ),
    ),
    factory.createStringLiteral(from),
  );
}
