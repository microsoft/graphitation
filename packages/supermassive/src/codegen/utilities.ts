import { factory } from "typescript";
import { DefinitionImport, TypeNameToTypeReference } from "./types";

export function createTypeNameToTypeReferenceMap(
  imports: DefinitionImport[],
): TypeNameToTypeReference {}
