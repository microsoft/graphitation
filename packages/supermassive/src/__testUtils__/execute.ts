import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  OperationDefinitionNode,
  parse,
  ExecutionArgs,
  ExecutionArgs as GraphQLExecutionArgs,
  ExecutionResult as GraphQLExecutionResult,
} from "graphql";
import { executeWithSchema } from "../executeWithSchema";
import { typeDefs } from "../benchmarks/swapi-schema";
import resolvers from "../benchmarks/swapi-schema/resolvers";
import { ExecutionResult, UserResolvers } from "../types";
import models from "../benchmarks/swapi-schema/models";
import { executeWithoutSchema } from "../executeWithoutSchema";
import { extractMinimalViableSchemaForRequestDocument } from "../utilities/extractMinimalViableSchemaForRequestDocument";
import { PromiseOrValue } from "graphql/jsutils/PromiseOrValue";
import { ObjMap } from "../jsutils/ObjMap";
import { forAwaitEach, isAsyncIterable } from "iterall";

type GraphQLResult<TData = ObjMap<unknown>, TExtensions = ObjMap<unknown>> =
  | GraphQLExecutionResult<TData, TExtensions>
  | AsyncGenerator<GraphQLExecutionResult<TData, TExtensions>, void, void>;

export function createExecutionUtils(
  graphqlExecute: (args: ExecutionArgs) => PromiseOrValue<ExecutionResult>,
  graphqlSubscribe: (
    args: ExecutionArgs,
  ) => PromiseOrValue<AsyncIterableIterator<ExecutionResult> | ExecutionResult>,
) {
  async function compareResultsForExecuteWithSchema(
    schema: GraphQLSchema,
    query: string,
    variables?: Record<string, unknown>,
  ) {
    expect.assertions(1);
    const document = parse(query);
    const result = await drainExecution(
      await executeWithSchema({
        document,
        definitions: typeDefs,
        resolvers: resolvers as UserResolvers<unknown, unknown>,
        contextValue: {
          models,
        },
        variableValues: variables,
      }),
    );
    const validResult = await drainExecution(
      await graphqlExecuteOrSubscribe({
        document,
        contextValue: {
          models,
        },
        schema,
        variableValues: variables,
      }),
    );
    expect(result).toEqual(validResult);
  }

  async function compareResultForExecuteWithoutSchemaWithMVSAnnotation(
    schema: GraphQLSchema,
    query: string,
    variables: Record<string, unknown> = {},
  ) {
    expect.assertions(1);
    const document = parse(query);
    const { definitions } = extractMinimalViableSchemaForRequestDocument(
      schema,
      document,
    );
    const result = await drainExecution(
      await executeWithoutSchema({
        document,
        contextValue: {
          models,
        },
        schemaFragment: {
          schemaId: "test",
          definitions,
          resolvers: resolvers as UserResolvers,
        },
        variableValues: variables,
      }),
    );
    const validResult = await drainExecution(
      await graphqlExecuteOrSubscribe({
        document,
        contextValue: {
          models,
        },
        schema,
        variableValues: variables,
      }),
    );
    expect(result).toEqual(validResult);
  }

  function graphqlExecuteOrSubscribe(
    args: GraphQLExecutionArgs,
  ): PromiseOrValue<GraphQLResult> {
    const operationName = args.operationName;
    let operation: OperationDefinitionNode | undefined;
    for (const definition of (args.document as unknown as DocumentNode)
      .definitions) {
      switch (definition.kind) {
        case Kind.OPERATION_DEFINITION:
          if (operationName == null) {
            if (operation !== undefined) {
              throw new Error("Bad operation in test");
            }
            operation = definition;
          } else if (definition.name?.value === operationName) {
            operation = definition;
          }
          break;
      }
    }
    if (!operation) {
      throw new Error("Bad operation in test");
    }

    if (operation.operation === "subscription") {
      return graphqlSubscribe(args) as GraphQLResult;
    } else {
      return graphqlExecute(args) as GraphQLResult;
    }
  }

  async function drainExecution(
    result: ExecutionResult | GraphQLResult,
  ): Promise<unknown> {
    let processedResult;
    if (isAsyncIterable(result)) {
      processedResult = await drainAsyncGeneratorToArray(result);
    } else if ("subsequentResults" in result) {
      processedResult = {
        ...result,
        subsequentResults: await drainAsyncGeneratorToArray(
          result.subsequentResults,
        ),
      };
    } else {
      processedResult = result;
    }
    return processedResult;
  }

  async function drainAsyncGeneratorToArray<T>(
    collection: AsyncGenerator<T, void, void>,
  ): Promise<T[]> {
    const result: T[] = [];
    await forAwaitEach(collection, (item) => result.push(item));
    return result;
  }

  return {
    compareResultForExecuteWithoutSchemaWithMVSAnnotation,
    compareResultsForExecuteWithSchema,
    drainExecution,
    graphqlExecuteOrSubscribe,
  };
}
