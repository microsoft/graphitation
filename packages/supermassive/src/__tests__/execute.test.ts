import {
  parse,
  execute as graphQLExecute,
  subscribe as graphQLSubscribe,
  GraphQLSchema,
} from "graphql";
import { makeSchema } from "../benchmarks/swapi-schema";
import models from "../benchmarks/swapi-schema/models";
import { createExecutionUtils } from "../__testUtils__/execute";
import { executeWithoutSchema } from "../executeWithoutSchema";
import { encodeASTSchema } from "../utilities/encodeASTSchema";
import { ResolveInfo } from "../types";

const {
  compareResultsForExecuteWithSchema,
  compareResultForExecuteWithoutSchemaWithMVSAnnotation,
  drainExecution,
  graphqlExecuteOrSubscribe,
} = createExecutionUtils(graphQLExecute, graphQLSubscribe);

interface TestCase {
  name: string;
  document: string;
  variables?: Record<string, unknown>;
}

const testCases: Array<TestCase> = [
  {
    name: "basic query",
    document: `
  {
    __typename
    person(id: 1) {
      name
      gender
    }
  }`,
  },

  {
    name: "include directive TRUE",
    document: `
  query Person($id: Int!, $include: Boolean!) {
    person(id: $id) {
      name
      gender @include(if: $include)
    }
  }
 `,
    variables: {
      id: 1,
      include: true,
    },
  },
  {
    name: "include directive FALSE",
    document: `
  query Person($id: Int!, $include: Boolean!) {
    person(id: $id) {
      name
      gender @include(if: $include)
    }
  }
 `,
    variables: {
      id: 1,
      include: false,
    },
  },
  {
    name: "skip directive TRUE",
    document: `
  query Person($id: Int!, $skip: Boolean!) {
    person(id: $id) {
      name
      gender @skip(if: $skip)
    }
  }
 `,
    variables: {
      id: 1,
      skip: true,
    },
  },
  {
    name: "skip directive FALSE",
    document: `
  query Person($id: Int!, $skip: Boolean!) {
    person(id: $id) {
      name
      gender @skip(if: $skip)
    }
  }
 `,
    variables: {
      id: 1,
      skip: false,
    },
  },
  {
    name: "basic query with variables",
    document: `
  query Person($id: Int!) {
    person(id: $id) {
      name
      gender
    }
  }
 `,
    variables: {
      id: 1,
    },
  },
  {
    name: "benchmark kitchen-sink query",
    document: `
    {
      empireHero: person(id: 1) {
        ...PersonFields
        height
        mass
      }
    
      jediHero: person(id: 4) {
        ...PersonFields
      }
    }
    
    fragment PersonFields on Person {
      name
      gender
      birth_year
      # TODO: separate field / arg for starships as async iterator (today it breaks with graphql-js >= 16)
      # starships {
      #  name
      # }
    }  
    `,
  },
  {
    name: "query with union",
    document: `
    {
      search(search: "Lu") {
        __typename
        ...on Person {
          name
        }
        ...on Planet {
          name
        }
        ... on Transport {
          name
        }
        ... on Species {
          name
        }
        ... on Vehicle{
          name
        }
      }
    }
    `,
  },
  {
    name: "query with interfaces",
    document: `
    {
      person: node(nodeType: Person, id: 1) {
        ... on Person {
          name
        }
      }
      vehicle: node(nodeType: Vehicle, id: 14) {
        ... on Vehicle {
          name
        }
      }
    }
    `,
  },
  {
    name: "with __typename",
    document: `
query Person($id: Int!) {
  person(id: $id) {
    __typename
  }
}
`,
    variables: {
      id: 1,
    },
  },
  {
    name: "union in interface fragment spread",
    document: `
    {
      person: node(nodeType: Person, id: 1) {
        ... on Alive {
          ... on Person {
            name
          }
        }
      }
    }
    `,
  },
  {
    name: "interface in union fragment spread",
    document: `
    {
      person: node(nodeType: Person, id: 1) {
        ... on Alive {
          ... on Node {
            id
          }
        }
      }
    }
    `,
  },
  {
    name: "interface in object fragment spread",
    document: `
    {
      person(id: 1) {
        ... on Node {
          id
        }
      }
    }
    `,
  },
  {
    name: "union in object fragment spread",
    document: `
    {
      person(id: 1) {
        ... on Alive {
          ... on Person {
            name
          }
        }
      }
    }
    `,
  },
  {
    name: "union in object fragment spread",
    document: `
    {
      person(id: 1) {
        ... on Vehicle {
          ... on Person {
            name
          }
        }
      }
    }
    `,
  },
  {
    name: "Default value in variables",
    document: `
    query ($title: String! = "The Empire Strikes Back") {
      searchFilmsByTitle(search: $title) {
        title
      }
    }`,
  },
  {
    name: "Default value",
    document: `
    {
      searchFilmsByTitle {
        title
      }
    }`,
  },
  {
    name: "Advanced Default value in variables",
    document: `
    query ($input: AdvancedInput! = { enumField: Film, otherField: "The Empire Strikes Back" }) {
      advancedDefaultInput(input: $input)
    }`,
  },
  {
    name: "Advanced Default value",
    document: `
    {
      advancedDefaultInput
    }`,
  },
  {
    name: "Not passing arg value vs passing arg value",
    document: `
    {
      multiArger
    }
    `,
  },
  {
    name: "Not passing arg value vs passing arg value",
    document: `
    {
      multiArger(a:null)
    }
    `,
  },
  // errors
  {
    name: "errors and nulls",
    document: `
    {
      person(id: 1) {
        name
        bubblingError
        bubblingListError
      }
    }
    `,
  },
  // subscription
  {
    name: "basic subscription with variables",
    document: `
  subscription emitPersons($limit: Int!) {
    emitPersons(limit: $limit) {
      name
      gender
    }
  }
 `,
    variables: {
      limit: 5,
    },
  },
  {
    name: "basic subscription with unused non-required variable",
    document: `
  subscription emitPersons($limit: Int!, $throwError: Boolean) {
    emitPersons(limit: $limit, throwError: $throwError) {
      name
      gender
    }
  }
 `,
    variables: {
      limit: 5,
    },
  },
  {
    name: "subscription throw an error",
    document: `
  subscription emitPersons($limit: Int!, $throwError: Boolean) {
    emitPersons(limit: $limit, throwError: $throwError) {
      name
      gender
    }
  }
 `,
    variables: {
      limit: 5,
      throwError: true,
    },
  },
  {
    name: "non-null query return null",
    document: `query { nonNullWithNull }`,
  },
  {
    name: "non-null subscription return null",
    document: `subscription { nonNullWithNull }`,
  },
  {
    name: "non-null query throw an error",
    document: `query { nonNullWithError }`,
  },
  {
    name: "non-null subscription throw in subscribe",
    document: `subscription { nonNullWithError }`,
  },
  {
    name: "non-null subscription throw in resolve",
    document: `subscription { nonNullWithErrorEvent }`,
  },
];

describe("graphql-js snapshot check to ensure test stability", () => {
  let schema: GraphQLSchema;
  beforeEach(() => {
    jest.resetAllMocks();
    schema = makeSchema();
  });

  test.each(testCases)("$name", async ({ document, variables }: TestCase) => {
    expect.assertions(1);
    const parsedDocument = parse(document);
    const result = await drainExecution(
      await graphqlExecuteOrSubscribe({
        schema,
        document: parsedDocument,
        contextValue: {
          models,
        },
        variableValues: variables,
      }),
    );
    expect(result).toMatchSnapshot();
  });
});

describe("executeWithSchema", () => {
  let schema: GraphQLSchema;
  beforeEach(() => {
    jest.resetAllMocks();
    schema = makeSchema();
  });

  test.each(testCases)("$name", async ({ document, variables }: TestCase) => {
    await compareResultsForExecuteWithSchema(schema, document, variables);
  });
});

describe("executeWithoutSchema - minimal viable schema annotation", () => {
  let schema: GraphQLSchema;
  beforeEach(() => {
    jest.resetAllMocks();
    schema = makeSchema();
  });
  test.each(testCases)("$name", async ({ document, variables }: TestCase) => {
    await compareResultForExecuteWithoutSchemaWithMVSAnnotation(
      schema,
      document,
      variables,
    );
  });
});

describe("executeWithoutSchema - regression tests", () => {
  test("Supports fieldNodes in ResolveInfo for backwards compatibility", async () => {
    let infoAtCallTime: ResolveInfo | undefined;

    const resolvers = {
      Query: {
        foo(_: unknown, __: unknown, ___: unknown, info: ResolveInfo) {
          infoAtCallTime = info;
        },
      },
    };
    const definitions = encodeASTSchema(parse("type Query { foo: String }"))[0];
    const document = parse(`{ foo, ... { foo } }`);

    await executeWithoutSchema({
      document,
      schemaFragment: { schemaId: "test", definitions, resolvers },
    });

    expect(infoAtCallTime?.fieldNodes?.length).toEqual(2);
    expect(infoAtCallTime?.fieldNodes[0].name.value).toEqual("foo");
    expect(infoAtCallTime?.fieldNodes[1].name.value).toEqual("foo");
  });
});
