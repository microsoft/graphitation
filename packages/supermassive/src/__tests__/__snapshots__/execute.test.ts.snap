// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-js snapshot check to ensure test stability Advanced Default value 1`] = `
{
  "data": {
    "advancedDefaultInput": "{"input":{"enumField":"Transport","otherField":"Foo"}}",
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability Advanced Default value in variables 1`] = `
{
  "data": {
    "advancedDefaultInput": "{"input":{"enumField":"Film","otherField":"The Empire Strikes Back"}}",
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability Default value 1`] = `
{
  "data": {
    "searchFilmsByTitle": [
      {
        "title": "The Empire Strikes Back",
      },
    ],
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability Default value in variables 1`] = `
{
  "data": {
    "searchFilmsByTitle": [
      {
        "title": "The Empire Strikes Back",
      },
    ],
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability Not passing arg value vs passing arg value 1`] = `
{
  "data": {
    "multiArger": "{}",
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability Not passing arg value vs passing arg value 2`] = `
{
  "data": {
    "multiArger": "{"a":null}",
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability basic query 1`] = `
{
  "data": {
    "__typename": "Query",
    "person": {
      "gender": "male",
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability basic query with variables 1`] = `
{
  "data": {
    "person": {
      "gender": "male",
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability basic subscription with unused non-required variable 1`] = `
[
  {
    "data": {
      "emitPersons": {
        "gender": "male",
        "name": "Luke Skywalker",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "n/a",
        "name": "C-3PO",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "n/a",
        "name": "R2-D2",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "male",
        "name": "Darth Vader",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "female",
        "name": "Leia Organa",
      },
    },
  },
]
`;

exports[`graphql-js snapshot check to ensure test stability basic subscription with variables 1`] = `
[
  {
    "data": {
      "emitPersons": {
        "gender": "male",
        "name": "Luke Skywalker",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "n/a",
        "name": "C-3PO",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "n/a",
        "name": "R2-D2",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "male",
        "name": "Darth Vader",
      },
    },
  },
  {
    "data": {
      "emitPersons": {
        "gender": "female",
        "name": "Leia Organa",
      },
    },
  },
]
`;

exports[`graphql-js snapshot check to ensure test stability benchmark kitchen-sink query 1`] = `
{
  "data": {
    "empireHero": {
      "birth_year": "19BBY",
      "gender": "male",
      "height": 172,
      "mass": 77,
      "name": "Luke Skywalker",
    },
    "jediHero": {
      "birth_year": "41.9BBY",
      "gender": "male",
      "name": "Darth Vader",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability errors and nulls 1`] = `
{
  "data": {
    "person": {
      "bubblingError": null,
      "bubblingListError": null,
      "name": "Luke Skywalker",
    },
  },
  "errors": [
    [GraphQLError: Bubbling!],
    [GraphQLError: Bubbling in list!],
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability include directive FALSE 1`] = `
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability include directive TRUE 1`] = `
{
  "data": {
    "person": {
      "gender": "male",
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability interface in object fragment spread 1`] = `
{
  "data": {
    "person": {
      "id": 1,
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability interface in union fragment spread 1`] = `
{
  "data": {
    "person": {
      "id": 1,
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability non-null query return null 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: Cannot return null for non-nullable field Query.nonNullWithNull.],
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability non-null query throw an error 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: Query error],
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability non-null subscription return null 1`] = `
[
  {
    "data": null,
    "errors": [
      [GraphQLError: Cannot return null for non-nullable field Subscription.nonNullWithNull.],
    ],
  },
]
`;

exports[`graphql-js snapshot check to ensure test stability non-null subscription throw in resolve 1`] = `
[
  {
    "data": null,
    "errors": [
      [GraphQLError: Subscription event error],
    ],
  },
]
`;

exports[`graphql-js snapshot check to ensure test stability non-null subscription throw in subscribe 1`] = `
{
  "errors": [
    [GraphQLError: Subscribe error],
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability query with interfaces 1`] = `
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
    },
    "vehicle": {
      "name": "Snowspeeder",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability query with union 1`] = `
{
  "data": {
    "search": [
      {
        "__typename": "Person",
        "name": "Luke Skywalker",
      },
      {
        "__typename": "Person",
        "name": "Luminara Unduli",
      },
      {
        "__typename": "Planet",
        "name": "Felucia",
      },
      {
        "__typename": "Planet",
        "name": "Sullust",
      },
      {
        "__typename": "Species",
        "name": "Sullustan",
      },
      {
        "__typename": "Transport",
        "name": "Raddaugh Gnasp fluttercraft",
      },
      {
        "__typename": "Vehicle",
        "name": "Raddaugh Gnasp fluttercraft",
      },
    ],
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability skip directive FALSE 1`] = `
{
  "data": {
    "person": {
      "gender": "male",
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability skip directive TRUE 1`] = `
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability subscription throw an error 1`] = `
{
  "errors": [
    [GraphQLError: error],
  ],
}
`;

exports[`graphql-js snapshot check to ensure test stability union in interface fragment spread 1`] = `
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability union in object fragment spread 1`] = `
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
    },
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability union in object fragment spread 2`] = `
{
  "data": {
    "person": {},
  },
}
`;

exports[`graphql-js snapshot check to ensure test stability with __typename 1`] = `
{
  "data": {
    "person": {
      "__typename": "Person",
    },
  },
}
`;
